String analysis with wordbag to generate dataset with work embeddings

In [34]:
import pandas as pd
import numpy as np

ori_df = pd.read_pickle("../dataset/features/raw_features.pickle")


In [35]:
ori_df[['Label', 'strings']].compute()

Label                                            strings
0        1  [!This program cannot be run in DOS mode., .rd...
1        1  [ARQCAAAAAADAAAAAAAAARg==, TAAAAA==, !This pro...
2        1  [!This program cannot be run in DOS mode., Ric...
3        1  [!This program cannot be run in DOS mode., `.r...
4        1  [!This program cannot be run in DOS mode., Ric...
..     ...                                                ...
316      0                                     [V3JpdGVGaWxl]
317      0  [, /4B5', y.a879, ~XDI$`, 6EI9{", 7B6 6, <&LU...
318      0  [V3JpdGVGaWxl, W[pe/O, ?Q*N(c, ceibLv, wm4t1r,...
319      0  [!This program cannot be run in DOS mode., P|R...
320      0  [ZItk, Vk1YaA==, V3JpdGVGaWxl, mi* >_, r}!!/z,...

[321 rows x 2 columns]

In [36]:
strs_df = ori_df[['Label', 'strings']].compute()

In [37]:
strs_df.head()["strings"][0]

['!This program cannot be run in DOS mode.',
 '.rdata',
 '`@.bss',
 '.edata',
 '0@.idata',
 '.reloc',
 '0B.gnu_deb',
 '\x1f<dVmT',
 '};\x1fTBT2pq',
 'P<)M,M',
 'PO@MA$6',
 'FOZM[$',
 "'W`ScS",
 'WgSVSc',
 'WtSCSt',
 '.M"OM/Z',
 'P7QJVu',
 'O*OIJBJ',
 '?"*~, ',
 '@QGizdL',
 'FLOTI}',
 'xAThFF',
 'c?(!".',
 "/'A(.9",
 '1ig8.QBE',
 '#+"p? ',
 'Vt9@ZpY',
 'cQri6lF]',
 'CwB8bq',
 'NiD:ht',
 '/@FHG4M',
 '__1NJK',
 '4sP\\.In',
 'pG9<dku',
 'X9JqN0',
 'krr&7>8>vZI9G',
 '\x1fAklw_E',
 'G$hG,F.',
 'X>\x1fJl\x1fwG',
 'K`\x1f+v/!m',
 '@A{C@K',
 '9Atc_fR',
 'GB;`!_\\',
 'VL8q\\O\x1f',
 'i6yvsxi',
 'pH9\x1f`\\',
 '_$>9Am',
 '`~^L>V',
 'G".,!y',
 'Bt&P":&y',
 "!!\\O^'>u",
 '&AY"!=',
 '"*KtR8',
 ')`"Kc2@u',
 ',($j"8(U',
 '<>9F:;5',
 '4aL):O',
 "aCD^')K",
 'Yyn~$q',
 'OABo."',
 ':AH?A_i',
 "v2Vn'K",
 'f?{J@oa',
 'L0O6Qh',
 'Q\x1fQ(Q*Q$[>Q',
 '@lT"?$',
 ',H4Gj8',
 'Kq2w9gw',
 '~R^AHe',
 "'Hbk\\_",
 'NR9K+9$B',
 ')\\\x1f(+`=',
 'FLH79!=',
 'IhL,[AL',
 'Bj(OdV',
 'KGCN,Ze',
 'LL?I9o',
 "*\\G9C'*.8",
 '7:?

In [38]:
import nltk
from nltk import FreqDist
from nltk.tokenize import word_tokenize, regexp_tokenize
import pandas as pd

nltk.download('punkt')

# define variables for wordbag
token_index = 1
wordbag = {}

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\ren\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [39]:
def tokens_gen(sen):
    return word_tokenize(sen)

def spe_tokens_gen(sen):
    if '\\' in sen:
        return sen.split("\\")
    elif ' ' not in sen:
        return sen
    else:    
        return word_tokenize(sen)

In [40]:
strs_df.head()["strings"][0]

['!This program cannot be run in DOS mode.',
 '.rdata',
 '`@.bss',
 '.edata',
 '0@.idata',
 '.reloc',
 '0B.gnu_deb',
 '\x1f<dVmT',
 '};\x1fTBT2pq',
 'P<)M,M',
 'PO@MA$6',
 'FOZM[$',
 "'W`ScS",
 'WgSVSc',
 'WtSCSt',
 '.M"OM/Z',
 'P7QJVu',
 'O*OIJBJ',
 '?"*~, ',
 '@QGizdL',
 'FLOTI}',
 'xAThFF',
 'c?(!".',
 "/'A(.9",
 '1ig8.QBE',
 '#+"p? ',
 'Vt9@ZpY',
 'cQri6lF]',
 'CwB8bq',
 'NiD:ht',
 '/@FHG4M',
 '__1NJK',
 '4sP\\.In',
 'pG9<dku',
 'X9JqN0',
 'krr&7>8>vZI9G',
 '\x1fAklw_E',
 'G$hG,F.',
 'X>\x1fJl\x1fwG',
 'K`\x1f+v/!m',
 '@A{C@K',
 '9Atc_fR',
 'GB;`!_\\',
 'VL8q\\O\x1f',
 'i6yvsxi',
 'pH9\x1f`\\',
 '_$>9Am',
 '`~^L>V',
 'G".,!y',
 'Bt&P":&y',
 "!!\\O^'>u",
 '&AY"!=',
 '"*KtR8',
 ')`"Kc2@u',
 ',($j"8(U',
 '<>9F:;5',
 '4aL):O',
 "aCD^')K",
 'Yyn~$q',
 'OABo."',
 ':AH?A_i',
 "v2Vn'K",
 'f?{J@oa',
 'L0O6Qh',
 'Q\x1fQ(Q*Q$[>Q',
 '@lT"?$',
 ',H4Gj8',
 'Kq2w9gw',
 '~R^AHe',
 "'Hbk\\_",
 'NR9K+9$B',
 ')\\\x1f(+`=',
 'FLH79!=',
 'IhL,[AL',
 'Bj(OdV',
 'KGCN,Ze',
 'LL?I9o',
 "*\\G9C'*.8",
 '7:?

In [41]:
def create_wordbag(sen_list):
    ''' create the wordbag models for url query parameters
    :param filename: the file with urls inside
    :param max: 
    '''
    tokens = []
    global wordbag
    global token_index
    for lines in sen_list:
        
        # for line in lines:
        #     if line is not None:
        #         # tokens.extend(tokens_gen(line))
        #         tokens.extend(spe_tokens_gen(line))

        line = ' '.join(lines)
        tokens.extend(word_tokenize(line))

    freddist = nltk.FreqDist(tokens)
    tokens = list(freddist.keys())
    for token in tokens:
        if token not in wordbag.keys():
            wordbag[token] = token_index
            token_index +=1
        else:
            continue

In [42]:
strings_list = strs_df['strings']

create_wordbag(strings_list)

In [43]:
def wordbag_vec(sens_list, labels_list):
    '''
    :param features_list: undecoded malicious payloads
    :param labels_list: label_encoded malicious types
    '''
    Y = []
    X = []
    X_lens = []
    for index, lines in enumerate(sens_list):
        # for line in lines:
        #     if isinstance(line, str):
        #         # tokens = tokens_gen(line)
        #         tokens = spe_tokens_gen(line)
        #         # define the vecs here
        #         vecs = []
        #         for token in tokens:
        #             if token in wordbag.keys():
        #                 vecs.append(wordbag[token])
        #             else:
        #                 vecs.append(-1)
                        
        #         X_lens.append(len(vecs))
        #         X.append(vecs)
        #         Y.append(labels_list[index])
    
        line = ' '.join(lines)
        tokens = tokens_gen(line)
        vecs = []
        for token in tokens:
            if token in wordbag.keys():
                vecs.append(wordbag[token])
            else:
                vecs.append(-1)
                
        X_lens.append(len(vecs))
        X.append(vecs)
        Y.append(labels_list[index])


    max_len = max(X_lens)
    for x in X:
        if len(x)!= max_len:
            x.extend((max_len-len(x))*[0])

    X = np.array(X)
    # return X, X_lens
    return X, np.array(Y)


In [44]:
labels_list = strs_df['Label']
X, Y = wordbag_vec(strings_list, labels_list)

In [48]:
print(X)
print(X.shape)
print(Y)
print(Y.shape)

from numpy import savetxt
savetxt("../dataset/matrix/X_str.csv", X)
savetxt("../dataset/matrix/Y_str.csv", Y)

[[4.37569270e-05 8.75138540e-05 1.31270781e-04 ... 0.00000000e+00
  0.00000000e+00 0.00000000e+00]
 [1.64760050e-02 1.64928001e-02 1.67951121e-05 ... 0.00000000e+00
  0.00000000e+00 0.00000000e+00]
 [5.85394370e-06 1.17078874e-05 1.75618311e-05 ... 0.00000000e+00
  0.00000000e+00 0.00000000e+00]
 ...
 [1.17562975e-02 6.17584100e-06 5.84201175e-06 ... 0.00000000e+00
  0.00000000e+00 0.00000000e+00]
 [3.19632561e-07 6.39265121e-07 9.58897682e-07 ... 0.00000000e+00
  0.00000000e+00 0.00000000e+00]
 [2.14747659e-03 2.33197921e-02 9.24587191e-03 ... 0.00000000e+00
  0.00000000e+00 0.00000000e+00]]
(321, 19524)
[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0
 0 0 0 0 0 0

In [45]:
# build binary classification model

from keras import Sequential
from keras import layers 
from keras.callbacks import EarlyStopping
def create_model(input_dim):
    model = Sequential()
    model.add(layers.Dense(32, input_dim=input_dim, activation='relu'))
    model.add(layers.Dense(32, activation='relu'))
    model.add(layers.Dense(1, activation='sigmoid'))
    return model

# Create model and train
model = create_model(X.shape[1])
model.compile(loss='binary_crossentropy', 
               optimizer='adam', 
               metrics=['accuracy'])

# normalization
from sklearn.preprocessing import Normalizer

transformer = Normalizer().fit(X)
X = transformer.transform(X)
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

history = model.fit(X_train, y_train, epochs=10, validation_split=0.2, shuffle=True, 
                batch_size=15, verbose=2, callbacks= [EarlyStopping('accuracy', patience=10)])


(321, 19524)
(321,)
Epoch 1/100
14/14 - 0s - loss: 0.6848 - accuracy: 0.5588 - val_loss: 0.6719 - val_accuracy: 0.5769
Epoch 2/100
14/14 - 0s - loss: 0.6425 - accuracy: 0.6176 - val_loss: 0.6481 - val_accuracy: 0.6346
Epoch 3/100
14/14 - 0s - loss: 0.5858 - accuracy: 0.7941 - val_loss: 0.6222 - val_accuracy: 0.6538
Epoch 4/100
14/14 - 0s - loss: 0.5117 - accuracy: 0.8578 - val_loss: 0.5996 - val_accuracy: 0.6923
Epoch 5/100
14/14 - 0s - loss: 0.4266 - accuracy: 0.9069 - val_loss: 0.5768 - val_accuracy: 0.6923
Epoch 6/100
14/14 - 0s - loss: 0.3404 - accuracy: 0.9363 - val_loss: 0.5810 - val_accuracy: 0.6923
Epoch 7/100
14/14 - 0s - loss: 0.2563 - accuracy: 0.9510 - val_loss: 0.5450 - val_accuracy: 0.7500
Epoch 8/100
14/14 - 0s - loss: 0.1879 - accuracy: 0.9804 - val_loss: 0.5693 - val_accuracy: 0.7308
Epoch 9/100
14/14 - 0s - loss: 0.1355 - accuracy: 0.9902 - val_loss: 0.5510 - val_accuracy: 0.7500
Epoch 10/100
14/14 - 0s - loss: 0.0978 - accuracy: 0.9902 - val_loss: 0.5784 - val_accura

In [46]:
# test the accuracy on test dataset
y_pre = model.predict_classes(X_test)
from sklearn.metrics import accuracy_score

accuracy_score(y_test, y_pre)

0.8

#### make the records with different settings

- Experiment1:
    - processing: tokens_gen 
    - accuracy: 0.59
    - label method: label every sen in one row
    - epochs: 10

---
- Experiment2:
    - processing: spe_tokens_gen: 
    - accuracy: 0.6
    - label method: label every sen in one row
    - epochss: 10

---
- Experiment3:
    - processing: tokens_gen 
    - accuracy: 0.99
    - label method: label joined strings
    - epochs: 10